Masonry backend

PoignardAzur opened the pull request:
[Rendered](https://github.com/linebender/rfcs/blob/10ecf78c617c762a0e606e82e2c35d14827d3540/rfcs/0001-masonry-backend.md ).

PoignardAzur committed:
Create masonry-backend RFC

PoignardAzur committed:
Remove placeholder file

DJMcNab commented on rfcs/0001-masonry-backend.md:
```diff
@@ -0,0 +1,166 @@
+- Creating a new `xilem_native` crate on parity with the current Xilem crate.
+- Removing the current xilem crate.
+- Publishing both `xilem_native` and the placeholder `xilem` crate on crates.io.
+- Sharing ownership of the crates with Raph Levien.
```
I'd like to avoid over-specialising on the release process, see e.g. [#linebender > Our Ideal Crate Release Process](https://xi.zulipchat.com/#narrow/stream/419691-linebender/topic/Our.20Ideal.20crate.20release.20processs).

Philipp-M replied:
Agree, the RFC feels like it should be multiple RFCs, as it's not just about masonry as a backend.

PoignardAzur replied:
Right. I'll move that part to "future possibilities".

DJMcNab commented on rfcs/0001-masonry-backend.md:
```diff
@@ -0,0 +1,166 @@
+
+### `xilem_native`
+
+The element type in `xilem_native` must implement `masonry::Widget`.
```
To confirm, `xilem_native` would not contain any `Widget` implementations?

DJMcNab commented on rfcs/0001-masonry-backend.md:
```diff
@@ -0,0 +1,166 @@
+- `xilem_web/`
+- `xilem_native/`
+- `masonry/`
+- `xilem/`, which would only be a placeholder crate pointing to `xilem_web/` or `xilem_native/`.
```
Pointing to here meaning:
a) A link in the readme, and no content
b) Re-exporting only `xilem_native`
c) Re-exporting `xilem_native` or `xilem_web`, depending on platform
?

DJMcNab replied:
There are tradeoffs here. My preference in the near term would be for `xilem` itself to be the crate which users depend on for the native backend, with no need to know about `xilem_native`.

However, that does constrain the longer-term web/native backend unification/abstraction, which I hope might happen, to not use `xilem` as the crate name, which would be unfortunate.

DJMcNab commented on rfcs/_placeholder.txt:
```diff
@@ -1 +0,0 @@
-Will be removed after first RFC is merged.
```
:fire:

DJMcNab reviewed (commented):
Overall, I think this is the right direction, but I'll need to give a more thorough review later.

Philipp-M commented on rfcs/0001-masonry-backend.md:
```diff
@@ -0,0 +1,166 @@
+- **The reactive layer** handles the app logic and centralized state. Your application's code will call functions from this layer.
+- **The platform layer** handles the document state, depending on the platform.
+
+The reactive layer defines the `View` trait as well as other items for sequences, type-erasure and memoization. A `View` represent a transient node or tree of graphical items, which is return by a declarative function we usually call a "component". An example of component could be:
```
```suggestion
The reactive layer defines the `View` trait as well as other items such as sequences of views, type-erasure and memoization. A `View` represents a transient description of one or multiple graphical items, which is usually returned by a declarative function that is often called a "component". An example of component could be:
```

Philipp-M replied:
Not really resolved, surely this is a little bit of nitpicking... but I don't see why the suggestion is not taken (as it should be a little bit more precise). And yes, I don't think the whole explanation of Xilems reactive architecture is necessary in this RFC (but since you've already written it, we can at least make it more accurate).

PoignardAzur replied:
> I don't see why the suggestion is not taken

The current text in the RFC:

> A `View` represents a transient description of one or multiple graphical items, which is usually returned by a declarative function that is often called a "component".

Is taken directly from your wording. I didn't keep the "such as sequences of views, type-erasure and memoization" part because the sentence "reactive layer defines [...] items such as [...] type-erasure and memoization" doesn't really parse right.

Philipp-M replied:
Well anyway, it's nitpicking, I'm fine with the current wording

Philipp-M commented on rfcs/0001-masonry-backend.md:
```diff
@@ -0,0 +1,166 @@
+
+In that example, `app_logic()` returns a `Column` view which contains two `Button` views. These store the minimum information to represent the concept of two buttons stored in a column.
+
+The `View` trait has an `Element` associated type, which depends on the platform layer, and represents a persisten UI entity. Every time app data changes, a new View is produced. That View is diffed against the app's previous View, and all changes are applied to matching elements: if the new view has a child view that wasn't in the previous version, a new element is added; if the new view no long ...
```
```suggestion
The `View` trait has an associated type `Element`, which depends on the platform layer, and represents a persistent UI entity. Every time app data changes, a new View is produced. That View is diffed against the app's previous View, and all changes are applied to matching elements: if the new view has a child view that wasn't in the previous version, a new element is added; if the new view no longer has a child view that was in the previous version, the matching element is removed, and so on.
```

Not sure whether it "depends on the platform layer" but is just the actual type of the UI Widget (The platform is rather defined by the runtime).
As you say yourself below, the `View` trait is a different type on each platform, requiring different `Widget` traits for the `Element` type
 
Might be nitpicking though...

PoignardAzur replied:
Yeah, I rushed that explanation a bit. Didn't want to have to document the entire xilem_core paradigm just to justify switching a backend.

Philipp-M commented on rfcs/0001-masonry-backend.md:
```diff
@@ -0,0 +1,166 @@
+
+## Drawbacks
+
+### Layer separation
+
+This move decouples the reactive layer of xilem from its platform layer by putting them in separate crates.
+
+This means some features in `xilem_native` may be harder to implement because they won't have access to the backend's private code.
```
I think this needs more detailed explanation, do you mean the whole runtime? Only the widgets? The xilem `Widget` trait?

I'm also not sure whether this should be delayed and just prototyped when deemed necessary (and based on that experience a new RFC or depending on the complexity just a PR opened), I don't see much benefit in splitting xilem in too many crates currently (mostly because of the orphan rule). AFAIU the widgets in masonry will be its own crate already anyway, so the likely biggest amount of code is already in a separate crate.

Philipp-M replied:
Not sure if this is really resolved is it? (same with the other comments) I don't see where this is explained in more detail in the new version. I can only guess, that the temporary `xilem_masonry` will contain the runtime (i.e. is the temporary merger) and this will be the new `xilem` at some point

PoignardAzur replied:
It depends on what you mean by "the whole runtime".

Broadly speaking, the RFC makes a distinction between two "layers", the reactive layer and the platform layer. These layers are fuzzy, but real: the main difference between xilem and xilem_web is that xilem_web's platform layer is built on top of the web DOM. The xilem crate also has `src/view/` and `src/widget/` that are symptoms of this separation too.

In xilem, the "platform layer" is basically "everything except the `view/` folder". This RFC proposes replacing all of that by masonry, which means what would be left would be the `view/` folder, plus glue code to interface with Masonry.

> I'm also not sure whether this should be delayed and just prototyped when deemed necessary (and based on that experience a new RFC or depending on the complexity just a PR opened), I don't see much benefit in splitting xilem in too many crates currently (mostly because of the orphan rule). AFAIU the widgets in masonry will be its own crate already anyway, so the likely biggest amount of code is already in a separate crate.

To be clear, in the native case, there would be only two crates: `xilem` and `masonry`. (Plus `xilem_core`, which defines macro that both `xilem` and `xilem_web` use).

I guess I should write a more complete explainer about this. It feels a bit annoying, because a lot of the complexity I'm being asked to explain and justify already exists in Xilem, but in any case documenting things is probably a good thing.

PoignardAzur replied:
I guess I could draw up a list of public items in `xilem` and split them between "frontend" and "backend". That would probably make the explanation a lot easier to grok.

Philipp-M replied:
Hmm reading again, it's *kinda* clear in the RFC (by defining reactive layer and platform layer), so sorry for being so persistent, but it's a bigger change...

You should probably include a lot what you've written here in the RFC itself, I think it clears up things a little bit.

Basically it boils down to this:
* Keep the reactive view layer (basically everything in `src/view/*`)  in xilem
* Move runtime and everything else into masonry (in this RFC platform layer)

I think we may want to have a third crate then that actually contains the runtime (`xilem` and another name for the reactive layer?). At least in the current form of xilem the separation between view <-> runtime is not that simple (as the runtime requires the `View` trait).

So that the dependency chain would look like this: runtime -> reactive/view -> widgets (masonry)

(Or keep the runtime + reactive layer in one crate, I think this would be the simplest way forward for now)

PoignardAzur replied:
Yeah, figuring out the composition root might be a bit tricky, but I don't think it should be included in the RFC.

> You should probably include a lot what you've written here in the RFC itself

Will do. I do need to make these things more explicit.

> So that the dependency chain would look like this: runtime -> reactive/view -> widgets (masonry)

Most of the "runtime" as it currently exists will be defined in Masonry. The rough idea is that, when you create a Masonry window, you give it a bunch of callbacks (or a trait impl or whatever) to be called on every paint, input event, etc. So the composition root of Xilem will be in the callbacks passed to Masonry's own composition root.

Philipp-M replied:
> Yeah, figuring out the composition root might be a bit tricky, but I don't think it should be included in the RFC.

I think this was the reason why things were unclear to me a little bit and thus the question (runtime vs reactive layer etc.), since it's interconnected above `widget/` and `view/` (mainly `App` and `AppTask`) and thus probably not cleanly separable between "platform" and "reactive" layer (or we have a different interpretation of "reactive" layer) with just 2 crates (`xilem` and `masonry`).

Sure a lot of the code in these modules can be laid out in masonry etc. but as you say details can be done in PRs (I think it will clear up when `xilem_masonry` is being implemented).

PoignardAzur replied:
I've added a new sub-section that mentions the source files and how they'll be affected, though what the composition root will look like is left a bit vague.

Let me know if you think it should be more clear/explicit!

Philipp-M replied:
Yeah I think continuing discussion makes more sense in the relevant parts of the RFC

Philipp-M commented on rfcs/0001-masonry-backend.md:
```diff
@@ -0,0 +1,166 @@
+
+More importantly, Masonry was built as a potential foundation for *multiple* native GUI crates, and it would be valuable to conserve that; therefore, while Masonry would still depend on crates from the Linebender ecosystem, it shouldn't be too tied to Xilem.
+
+### Porting Xilem to Masonry vs Masonry to Xilem
+
+In any merge between two projects, one has to be the base the other is grafted on.
+
+This RFC was written under the assumption Xilem would move towards Masonry, but the opposite could be possible.
+
+That said, the Motivation section explains why Masonry is a healthier codebase to start from.
```
I expect to just take the best code in one or the other and just use that (and otherwise merge the good bits of the other).
Xilem is the frontfacing layer, so it depends on masonry otherwise?

PoignardAzur replied:
Xilem as it currently is has both frontend and backend code. The question how much of that backend code we want to build from vs throwing it away and starting from Masonry's code.

Philipp-M replied:
Yes and as I've written, I think the good parts can just be merged, I don't think it's an either/or thing

Philipp-M commented on rfcs/0001-masonry-backend.md:
```diff
@@ -0,0 +1,166 @@
+
+## Future possibilities
+
+Xilem has added features over Druid that Masonry doesn't support, notably virtual lists and async support.
```
Well they're prototypes, and as you've already said, they are commented out, and quite a bit out of date/not-working (and I think there may be some possible UX improvements, such as the list callback taking the app state as well as parameter). I think when we're approaching virtualized lists etc. that there will be a refactored version implemented either in xilem as well as a corresponding widget in masonry. A working version is currently on the old [`xilem_tokio`](https://github.com/linebender/druid/tree/xilem_tokio) branch in druid (which is basically the same code as currently in xilem, but working...)

async is an interesting point though, I think long-term async will almost be a requirement for some widgets in masonry. Possibly via a feature gate for async widgets (ideally not hardcoding a specific async runtime)?

PoignardAzur committed:
Move out-of-scope items to "Future possibilities" section

PoignardAzur committed:
Rewrite some sections for clarity.

fix small typos.

PoignardAzur commented:
RFC has been updated to reflect feedback; plus I reworded a few things, fixed typos, etc.

giannissc commented:
I am not sure what level of detail you want to see on the RFC, but I was wondering in which ways are masonry and Xilem (backend) different, and what are the tradeoffs of each approach? I am thinking specifically about the traits here, but there might be other significant differences that might be worth discussing.

Philipp-M commented on rfcs/0001-masonry-backend.md:
```diff
@@ -0,0 +1,186 @@
+  - (soon) `masonry/`
+- `src/`
+
+In other words, all crates are in a `crates/` directory, except for `xilem/` which is in the root. This makes it seem like eg `xilem_web` is a dependency of `xilem`, when they're disjoint libraries.
```
We could also move everything to `crates/` (so `crates/xilem`) and make it a pure workspace. Though I don't think the current structure is too bad, especially given that `xilem_web` is still kinda experimental (and either diverges fully, in case a separate repo may make sense, or can be hopefully merged in one way or the other at some point)

Relevant: https://xi.zulipchat.com/#narrow/stream/354396-xilem/topic/Moving.20.28or.20splitting.29.20the.20xilem.20crate.3F

raphlinus commented on rfcs/0001-masonry-backend.md:
```diff
@@ -0,0 +1,186 @@
+
+However, these changes were deemed out of scope for this RFC.
+
+### Ownership story
```
I think it would be worthwhile to mention that parley is in very much the same state - an individual project that we want to bring under the linebender umbrella.

jplatte commented:
I know my opinion isn't all that important since I haven't contributed much but I like the current proposal. No concerns :)

PoignardAzur committed:
Update the RFC to account for feedback.

Explain the source-code-level impact of the RFC.
Mention parley in "Ownership story".

PoignardAzur commented:
RFC has been updated to account for feedback. I think it's ready for review, though realistically there's probably stuff that could still be added (eg I'm now realizing it's very light on code snippets).

Philipp-M commented on rfcs/0001-masonry-backend.md:
```diff
@@ -0,0 +1,211 @@
+
+All of this source code except for the `view/` folder and `app.rs` will be removed.
+
+Xilem will still export a view trait, defined in `src/view/View.rs` using macros from `xilem_core`. Views will be rewritten to have `masonry::Widget` as their `Element` associated type.
```
```suggestion
Xilem will still export a view trait, defined in `src/view/view.rs` using macros from `xilem_core`. Views will be rewritten to have `masonry::Widget` as their `Element` associated type.
```

Philipp-M commented on rfcs/0001-masonry-backend.md:
```diff
@@ -0,0 +1,211 @@
+
+Xilem will still export a view trait, defined in `src/view/View.rs` using macros from `xilem_core`. Views will be rewritten to have `masonry::Widget` as their `Element` associated type.
+
+The `app.rs` composition root will be rewritten to be a minimal wrapper around Masonry's composition root. The Masonry composition root essentially takes callbacks when being built; these callbacks will be piped to the Xilem reactive architecture (building a view tree, reconciliation, and events).
```
I'm not sure I would go into that detail (callbacks etc.). I think this will clarify in PRs etc. (Because I don't think it will end up just as that, but we'll see).

I think it's enough to just clarify that `app.rs` or `App` (or something along the lines) takes the tasks of building the view tree, reconciliation, and (towards-views) events, and the rest is basically done by masonry which is owned by `app.rs` (however these things communicate, and in whichever threads they may live).

Philipp-M reviewed (commented):
In the risk of getting stuck in details (but discussion/clarification is probably not a bad thing anyway, so that more informed decisions can be made):

How do you think will the Widget (masonry) layer communicate with the View layer? Currently this is done via `IdPath` in xilem (that would be more precisely within `Switch` and `Button`).
Do you plan to port the `IdPath` into Masonry widgets?

xStrom commented:
With Druid we had `druid-shell` as a separate crate but in the same repo. It definitely made development easier, but also worked against the message that other toolkits could use it as well.

Then when we forked it to Glazier, it was moved to its own repository. To make the separation from Xilem more clear. I wonder, why doesn't that rationale repeat here for Masonry? Wouldn't we be repeating the `druid-shell` mistake where people would either be unaware of Masonry or think that it is just some bespoke cog of Xilem?

I do think that if we want to be serious about the goal of Masonry being for more than just for Xilem, then it should be in its own repository. (*But moved to under Linebender*) It would send a much stronger signal that Masonry stands without Xilem. Also, just in terms of issue management and noise. Do we really want non-Xilem Masonry issues, i.e. issues affecting other consumers of Masonry, present in the Xilem repo?

Some additional coordination would definitely be required for the separate repo approach, however we've done it before with Piet and we still do it with Vello, Kurbo, etc. It's not that insurmountable.

If however we want to be much more focused on Xilem's needs and not really worry about Masonry succeeding beyond Xilem, well then it might as well be a module or even just code inside existing modules. It would allow for deeper integration and code sharing beyond what is possible across crates. It would cut down on clerical work around publishing, and more importantly cut all the maintenance work of supporting non-Xilem consumers of Masonry.

The part of the rationale for having it be a separate crate inside the Xilem repo that it would keep us honest for potential non-Xilem consumers seems like a big YAGNI to me. We should either deeply integrate it, effectively assimilating Masonry. Or if we pay the cost of catering to other toolkits, then be more serious about it.
